src/Data/Map.purs

```diff
@@ -104,6 +109,35 @@ instance traversableMap :: Traversable (Map k) where
           <*> traverse f right
   sequence = traverse id
 
+-- | The Semigroup instance combines values via a `Newtype w v`.
+newtype OnValues w k v = OnValues (Map k v)
+
+onValues :: forall w k v. Newtype w v => (v -> w) -> Map k v -> OnValues w k v
+onValues _ = OnValues
+
+derive instance newtypeOnValues :: Newtype (OnValues w k v) _
+derive newtype instance eq1OnValues :: Eq k => Eq1 (OnValues w k)
+derive newtype instance eqOnValues :: (Eq k, Eq v) => Eq (OnValues w k v)
+derive newtype instance ordOnValues :: (Ord k, Ord v) => Ord (OnValues w k v)
+derive newtype instance ord1OnValues :: Ord k => Ord1 (OnValues w k)
+derive newtype instance showOnValues :: (Show k, Show v) => Show (OnValues w k v)
+derive newtype instance functorOnValues :: Functor (OnValues w k)
+derive newtype instance foldableOnValues :: Foldable (OnValues w k)
+derive newtype instance traversableOnValues :: Traversable (OnValues w k)
+
+appendOnValues :: forall w k v.
+  Ord k =>
+  Semigroup w =>
+  Newtype w v =>
+  OnValues w k v -> OnValues w k v -> OnValues w k v
+appendOnValues (OnValues l) (OnValues r) = OnValues $ unionWith (\a b -> unwrap (wrap a <> wrap b :: w)) l r
+
+instance semigroupOnValues :: (Ord k, Semigroup w, Newtype w v) => Semigroup (OnValues w k v) where
```

Comment:
Newtyping the values seems odd to me, since the user could always do the newtype unwrapping. Why not just have

```
(Ord k, Semigroup v) => Semigroup (OnValues k v)
```
?